internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -460,6 +470,136 @@ func (r *BMACReconciler) findBMH(ctx context.Context, agent *adiiov1alpha1.Agent
+	if err != nil {
+		r.Log.WithError(err).Errorf("failed to create or update spoke Machine")
+		return reconcileError{err}
+	}
+
+	_, err = r.ensureSpokeBMH(ctx, spokeClient, bmh, clusterDeployment, machine)
+	if err != nil {
+		r.Log.WithError(err).Errorf("failed to create or update spoke BareMetalHost")
+		return reconcileError{err}
+	}
+
+	return reconcileComplete{}
+}
+
+func (r *BMACReconciler) ensureSpokeBMH(ctx context.Context, spokeClient client.Client, bmh *bmh_v1alpha1.BareMetalHost, clusterDeployment *hivev1.ClusterDeployment, machine *machinev1beta1.Machine) (*bmh_v1alpha1.BareMetalHost, error) {
+	bmhSpoke := &bmh_v1alpha1.BareMetalHost{}
+	err := spokeClient.Get(ctx, types.NamespacedName{Name: bmh.Name, Namespace: bmh.Namespace}, bmhSpoke)
+	if err != nil && errors.IsNotFound(err) {
+		bmhSpoke = &bmh_v1alpha1.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      bmh.Name,
+				Namespace: bmh.Namespace,
+			},
+		}
+		bmhSpoke.Spec = bmh.Spec
+		bmhSpoke.Spec.ExternallyProvisioned = true
+		bmhSpoke.Spec.ConsumerRef.Name = machine.Name
+		bmhSpoke.Spec.ConsumerRef.Namespace = machine.Namespace
+		r.Log.Infof("Creating spoke BareMetalHost %s/%s for cluster %s", bmhSpoke.Namespace, bmhSpoke.Name, clusterDeployment.Name)
+		err = spokeClient.Create(ctx, bmhSpoke)
+		if err != nil {
+			r.Log.WithError(err).Errorf("failed to create spoke BareMetalHost %s/%s", bmhSpoke.Namespace, bmhSpoke.Name)
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	}
+	updated := bmh.DeepCopy()
+	updated.Spec.ExternallyProvisioned = true
+	updated.Spec.ConsumerRef.Name = machine.Name
+	updated.Spec.ConsumerRef.Namespace = machine.Namespace
+	if !reflect.DeepEqual(updated.Spec, bmhSpoke.Spec) {
+		err = spokeClient.Update(ctx, updated)
+		if err != nil {
+			r.Log.WithError(err).Errorf("failed to update spoke BareMetalHost %s/%s", updated.Namespace, updated.Name)
+			return nil, err
+		}
+	}
+	return bmhSpoke, nil
+}
+
+func (r *BMACReconciler) ensureSpokeMachine(ctx context.Context, spokeClient client.Client, bmh *bmh_v1alpha1.BareMetalHost, clusterDeployment *hivev1.ClusterDeployment) (*machinev1beta1.Machine, error) {
+	machine := &machinev1beta1.Machine{}
+	machineName := fmt.Sprintf("%s-%s", clusterDeployment.Name, bmh.Name)
+	err := spokeClient.Get(ctx, types.NamespacedName{Name: machineName, Namespace: bmh.Namespace}, machine)
+	if err != nil && errors.IsNotFound(err) {
+		machine = &machinev1beta1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      machineName,
+				Namespace: bmh.Namespace,
+			},
+		}
+		machine.Labels = r.setMachineLabels(machine.Labels, clusterDeployment)
+		r.Log.Infof("Creating spoke Machine %s/%s for cluster %s", machine.Namespace, machine.Name, clusterDeployment.Name)
+		err = spokeClient.Create(ctx, machine)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, err
+	}
+
+	machine.Labels = r.setMachineLabels(machine.Labels, clusterDeployment)
```

Comment:
still the same question. If we have nothing to update why we should?

Reply:
Added a check for if we need to update.